tests/access-control.tests.ts

```diff
@@ -0,0 +1,257 @@
+import { consts, Roles } from './constants'
+import {bnArg, expect, fromSigner, setupContract} from './helpers'
+
+describe('MY_ACCESS_CONTROL', () => {
+  async function setup() {
+    return setupContract('my_access_control', 'new')
+  }
+
+  it('ERC 721 - mint works', async () => {
+    const {
+      contract,
+      query,
+      defaultSigner: sender
+    } = await setup()
+
+    await expect(contract.tx.mint(bnArg(0)))
+      .to.emit(contract, 'Transfer').withArgs(consts.EMPTY_ADDRESS, sender.address, bnArg(0))
+
+    const result = await query.balanceOf(sender.address)
+    expect(result.output).to.equal(1)
+  })
+
+  it('ERC 721 - mint existing should fail', async () => {
+    const {
+      contract,
+      query,
+      defaultSigner: sender
+    } = await setup()
+
+    await expect(contract.tx.mint(bnArg(0)))
+      .to.emit(contract, 'Transfer').withArgs(consts.EMPTY_ADDRESS, sender.address, bnArg(0))
+
+    const balance = await query.balanceOf(sender.address)
+    expect(balance.output).to.equal(1)
+
+    const owner = await query.ownerOf(bnArg(0))
+    expect(owner.output).to.equal(sender.address)
+
+    await expect(contract.tx.mint(bnArg(0))).to.eventually.be.rejected
+  })
+
+  it('ERC 721 - approved transfer works', async () => {
+    const {
+      contract,
+      query,
+      defaultSigner: sender,
+      accounts: [alice, bob]
+    } = await setup()
+
+    await expect(contract.tx.mint(bnArg(0)))
+      .to.emit(contract, 'Transfer').withArgs(consts.EMPTY_ADDRESS, sender.address, bnArg(0))
+
+    const owner = await query.ownerOf(bnArg(0))
+    expect(owner.output).to.equal(sender.address)
+
+    await expect(contract.tx.approve(alice.address, bnArg(0)))
+      .to.emit(contract, 'Approval').withArgs(sender.address, alice.address, bnArg(0))
+
+    await expect(fromSigner(contract, alice.address).tx.transferFrom(sender.address, bob.address, bnArg(0)))
+      .to.emit(contract, 'Transfer').withArgs(sender.address, bob.address, bnArg(0))
+
+    const ownerOf = await query.ownerOf(bnArg(0))
+    expect(ownerOf.output).to.equal(bob.address)
```

Comment:
You can create to.have.output(...) matcher and do all checks in one row:
```
chai.use((chai) => {
  chai.Assertion.addMethod('output', async function (param, message) {
    await new chai.Assertion(this._obj).to.eventually.have.property('output').to.equal(param, message)
  })
})
export const { expect } = chai
```

Reply:
Good point. thanks